### PR TITLE
📖 Document breaking changes policy

### LINKED ISSUE #1598: Shifting dependency versions in same release branch causing downstream issues
**What steps did you take and what happened:**
There seems to be some module weirdness originating out of the [Cluster API Upgrade Tool](https://github.com/vmware/cluster-api-upgrade-tool/blob/master/go.mod) centered around the [`github.com/Azure/go-autorest`](https://github.com/Azure/go-autorest) package. 

The upgrade tool pulls in `k8s.io/client-go@v11.0.1-0.20190409021438-1a26190bd76a+incompatible` which does *not* import autorest. However, autorest shows up several times as a transitive dependency for the upgrade tool:

```shell
$ cd $(mktemp -d) && \
  git clone https://github.com/vmware/cluster-api-upgrade-tool . && \
  go list -m -u all 2>/dev/null | grep autorest
github.com/Azure/go-autorest/autorest v0.9.0 [v0.9.2]
github.com/Azure/go-autorest/autorest/adal v0.6.0 [v0.8.0]
github.com/Azure/go-autorest/autorest/date v0.2.0
github.com/Azure/go-autorest/autorest/mocks v0.3.0
github.com/Azure/go-autorest/logger v0.1.0
github.com/Azure/go-autorest/tracing v0.5.0
```

The upgrade tool is inheriting the transitive dependency on autorest via the direct dependency on `sigs.k8s.io/cluster-api@v0.2.1`:

```shell
$ cd $(mktemp -d) && \
  git clone https://github.com/kubernetes-sigs/cluster-api . && \
  git checkout -b v0.2.1 v0.2.1 && \
  go list -m -u all 2>/dev/null | grep autorest
github.com/Azure/go-autorest/autorest v0.9.0 [v0.9.2]
github.com/Azure/go-autorest/autorest/adal v0.6.0 [v0.8.0]
github.com/Azure/go-autorest/autorest/date v0.2.0
github.com/Azure/go-autorest/autorest/mocks v0.3.0
github.com/Azure/go-autorest/logger v0.1.0
github.com/Azure/go-autorest/tracing v0.5.0
```

However, what happens when a downstream project imports both the upgrade tool *and* a newer version of Cluster API v1alpha2?

```shell
$ cd $(mktemp -d) && \
  git clone https://github.com/kubernetes-sigs/cluster-api . && \
  git checkout -b v0.2.5 v0.2.5 && \
  go list -m -u all 2>/dev/null | grep autorest
github.com/Azure/go-autorest v11.1.2+incompatible [v13.2.0+incompatible]
```

Conflict!

FWIW, the dependency on `github.com/Azure/go-autorest` is transitive via `k8s.io/client-go`.

Essentially, because Cluster API moves so quickly, and we're starting to see libraries, such as the upgrade tool, built on top of Cluster API, dependency conflicts are being created when downstream work wants to use both the most recent patch of Cluster API with a library built on top of Cluster API that still references an older version from the same release branch.

It makes me wonder if Cluster API should stop updating dependencies unless otherwise absolutely required during a release branch? 

* Cluster API v0.2.1 depends on `k8s.io/client-go v11.0.1-0.20190409021438-1a26190bd76a+incompatible`
* Cluster API v0.2.5 depends on `k8s.io/client-go v0.0.0-20190918200256-06eb1244587a`

The result is a messy situation for people trying to use Cluster API and libraries built on top of Cluster API.

One thing we could do to combat this is to start adding comments to `go.mod` files such as:

```
// The following dependency was resolved from the tag "kubernetes-1.15.4"
k8s.io/client-go v0.0.0-20190918200256-06eb1244587a
```

With anything related to k/k, the above comment approach would simplify aligning all k/k related dependencies in a project that imports both Cluster API and libraries built on top of it.

**What did you expect to happen:**
A downstream consumer of Cluster API should be able to pull in a newer version of Cluster API from the same release branch, even when another dependency pulls in a different version of Cluster API from the same release branch.

cc @ncdc

### PR DESCRIPTION
**Pull request by ncdc**
**What this PR does / why we need it**:
Document breaking changes guidelines in the contributing doc. Also clean
up word wrapping.

Note, most of the diffs are just word wrapping the paragraphs, with the addition of the breaking changes section. I also cleaned up the formatting of the list with the PR title emojis.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #1598

### CODE REVIEW

**Comment by ncdc** [comment]
cc @timothysc

File: CONTRIBUTING.md [hunk 1]
```diff
@@ -19,53 +19,91 @@ Read the following guide if you're interested in contributing to cluster-api.
...
-Some release branches might ship with both Go modules and dep (e.g. `release-0.1`), users backporting patches should always make sure
-that the vendored Go modules dependencies match the Gopkg.lock and Gopkg.toml ones by running `dep ensure`
+Cluster API maintains older versions through `release-X.Y` branches. We accept backports of bug fixes to the most recent
+release branch. For example, if the most recent branch is `release-0.2`, and the `master` branch is under active
+development for v0.3.0, a bug fix that merged to `master` that also affects `v0.2.x` may be considered for backporting
+to `release-0.2`. We generally do not accept PRs against older release branches.
+
+## Breaking Changes
+
+Breaking changes are generally allowed in the `master` branch, as this is the branch used to develop the next minor
```
Comment (vincepri) [inline CONTRIBUTING.md]: We should specify that this applies only if master is open for the next iteration

File: CONTRIBUTING.md [hunk 2]
```diff
@@ -19,53 +19,91 @@ Read the following guide if you're interested in contributing to cluster-api.
...
+
+There may, at times, need to be exceptions where breaking changes are allowed in release branches. These are at the
+discretion of the project's maintainers, and must be carefully considered before merging. An example of an allowed
+breaking change might be a fix for a behavioral bug that was released in an initial minor version (such as `v0.3.0`).
 
 ### Merge Approval
 
-Cluster API maintainers may add "LGTM" (Looks Good To Me) or an equivalent comment to indicate that a PR is acceptable. Any change requires at least one LGTM.  No pull requests can be merged until at least one Cluster API maintainer signs off with an LGTM.
+Cluster API maintainers may add "LGTM" (Looks Good To Me) or an equivalent comment to indicate that a PR is acceptable.
+Any change requires at least one LGTM.  No pull requests can be merged until at least one Cluster API maintainer signs
```
Comment (vincepri) [inline CONTRIBUTING.md]: This is hard to do in practice, given that any k8s contributor can, in theory add /lgtm to anything

Comment (ncdc) [reply CONTRIBUTING.md]: This isn't net new but I can adjust it

File: CONTRIBUTING.md (hunk 1 above)
Comment (ncdc) [reply CONTRIBUTING.md]: I know Kubernetes sends out updates throughout the release cycle informing about the state of the master branch. Maybe we should do something similar (defer to #768?)

Comment (vincepri) [reply CONTRIBUTING.md]: That sounds good to me, let's document that as well?

Comment (ncdc) [reply CONTRIBUTING.md]: Yes, but as part of 768

Comment (ncdc) [reply CONTRIBUTING.md]: Eek, I linked to the wrong thing. Nevermind about 768 😄. We had talked at the f2f about formalizing the release cycle, which is what I was trying to link to, but I don't think we ever created an issue for it (or I can't find it quickly).

Comment (ncdc) [reply CONTRIBUTING.md]: Updated PTAL

**Pull request merged**